src/main/java/org/terasology/alterationEffects/regenerate/RegenerationAlterationEffect.java

```diff
@@ -1,138 +1,74 @@
-            delayManager.addDelayedAction(entity, AlterationEffects.EXPIRE_TRIGGER_PREFIX + AlterationEffects.REGENERATION, duration);
-        }
-        // Otherwise, if there are either no modifiers found, or none of the modifiers collected in the event have infinite
-        // duration, remove the component associated with this regeneration effect.
-        else if (!modifiersFound || !effectModifyEvent.getHasInfDuration()) {
-            entity.removeComponent(RegenerationComponent.class);
-        }
-        // If this point is reached and none of the above if-clauses were met, that means there was at least one modifier
-        // collected in the event which has infinite duration.
+    @Override
+    protected RegenerationComponent updateComponent(OnEffectModifyEvent event, RegenerationComponent component,
+                                                    final EffectContext context) {
+        component.regenerationAmount = (int) event.getMagnitudeResultValue();
+        return component;
     }
 
     /**
-     * This will apply the regeneration effect on the given entity by calling the method
-     * {@link #applyEffect(EntityRef, EntityRef, float, long)}.
+     * This will apply the regeneration effect on the given entity by calling the method {@link #applyEffect(EntityRef,
+     * EntityRef, float, long)}.
      *
-     * @param instigator    The entity who applied the regen effect.
-     * @param entity        The entity that the regen effect is being applied on.
-     * @param id            Id is used for unique identification in regen scheduler.
-     * @param magnitude     The magnitude of the regen effect.
-     * @param duration      The duration of the regen effect.
+     * @param instigator The entity who applied the regen effect.
+     * @param entity The entity that the regen effect is being applied on.
+     * @param id Id is used for unique identification in regen scheduler.
+     * @param magnitude The magnitude of the regen effect.
+     * @param duration The duration of the regen effect.
      */
     @Override
     public void applyEffect(EntityRef instigator, EntityRef entity, String id, float magnitude, long duration) {
+        //TODO: this is supposed to be the "new" implementation with the approach above being deprecated
         //applyEffect(instigator, entity, magnitude, duration);
         if (magnitude != 0) {
+            //FIXME(skaldarnar): "duration" may be -1 to indicated DURATION_INDEFINITE - how does that translate to the
+            //                   regen event we are sending here?
+            //                   Actually, -1 has a special meaning for the regen effect, but it is different from the
+            //                   semantic in alteration effects: heal until the entity has restored full health as
+            //                   opposed to heal indefinitely
             entity.send(new ActivateRegenEvent(id, magnitude, ((float) duration) / 1000));
```

Comment:
I think you can send a new ActivateRegenEvent with duration -1 every time the ActivateRegenEvent with BaseRegen is sent (Although I think that is still not a good solution fora situation when for some reason baseRegen is disabled and potions are used..) so maybe we could send a new ActivateRegenEvent when.. doDamageEvent is sent which is the only time when ActivateRegen() is currently sent. 
This would probably mean having a separate logic for an indefinite duration, although I'm not really sure.... 

You could alter the healthComponent's regenRate which would mean modifying the base regen which would probably fix the duration issue but then again there's the situation where base regen is disabled..

You could maybe send a new AbstractValueModifiableEvent just before regen but that would mean modifying the whole current system and then again the base regen thing..

So I think sending a new ActivateRegen whenever current baseRegen is sent(i.e. when DoDamageEvent is sent) would take care of these problems but I don't know if that's something we'd like to do...

I don't really know how this can be done.. just wanted to write what I think about this even if it isn't much help 😅

Reply:
Thanks for your thoughts - I'm pretty sure it will be of help when actually reworking how the effects work :nerd_face: Not addressing it this PR though.